### DISCUSSION TITLE
Improve Undo, persist history buffer across tab closure

### DISCUSSION THREAD

**Discussion by 1hko** [zed-industries/zed#16485] [Feature Requests]
VSCode makes me fearless closing files amongst my crowded tabs, because I know I will never lose work or that work's context, _including_ that work's recent changes. In Zed, when I close a file, all of that history is gone :(

1. Make a change in a file
2. Save it
3. Close the file
4. Reopen the file using <kbd>command</kbd>-<kbd>shift</kbd>-<kbd>T</kbd>
5. Press <kbd>command</kbd>-<kbd>Z</kbd> to undo.. nothing to undo 👎 

It I close an entire project window or quit Zed, I think it's safe to clear file history. That's how VSCode handles it anyway. In any other case, maintaining this valuable history has become essential to my workflow, freely opening and closing files without losing any work.

**Comment by towhans** [comment]
I store undo history per file in vim. Simple and powerful. I miss that feature in Zed.
```
" Keep undo history across sessions by storing it in a file
if has('persistent_undo')
    let myUndoDir = expand(vimDir . '/undodir')
    " Create dirs
    call system('mkdir ' . vimDir)
    call system('mkdir ' . myUndoDir)
    let &undodir = myUndoDir
    set undofile
endif
```

**Reply by nixpulvis** [reply]
There is a minor security concern with persistent undo history, but I think it's still very important, and I was surprised it wasn't supported out of the box here.

**Comment by tintor** [comment]
Lack of persistent undo with auto-save is the main thing preventing me from moving from vim to zed.

Persistent undo history that is persisted even if zed is closed.
It is ok to clear versions from undo history older than a month.

**Comment by viniciusnevescosta** [comment]
I wanted to share a very stressful situation. I just lost an entire file of code because I right clicked on another file name. For some reason the editor selects 2 files (the one I was editing and the one I selected), and this behavior doesn't make sense, if I wanted to select 2 files I would have simply pressed shift or cmd and then clicked on the files. I ended up losing an empty file that I wanted and what I was editing. I'm sorry, but this should be a basic function of an IDE.

**Comment by maciej-lech** [comment]
Yes, after few weeks of trying zed, it lacks basic functionalities and looks like development focus more on AI stuff. Shame, as the whole editor is pretty fast, so it would be nice replacement of overgrown IDEs :/

**Comment by rdvo** [comment]
is there a way to see a timeline of recent edits like in vscode? or just undo

**Comment by lcrownover** [comment]
Just wanted to bump this. I *love* the ability in vim to open a file I was working on yesterday and be able to undo changes as though I never closed the file.

**Comment by numbereddev** [comment]
I can only bump this, PLEASE

**Comment by RWPXT** [comment]
Why isn't this a thing yet? This seems crucial to an IDE no? If I close a file and reopen a file, I would expect to be able to cmd+z. It's actually blowing my mind that I cannot do this.

**Comment by abhirammkaushik** [comment]
Can we bump this up, please? It's a frustrating experience for sure!

**Comment by rotarur** [comment]
me too, missing this feature in zed. bumping it up

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links

**Reply by nixpulvis** [reply]
Discord sucks for finding resources online. I strongly urge you to change this policy.

**Comment by Sekara** [comment]
I just accidentally lost some important information because I accidentally saved the file and closed Zed. Opening Zed and Cmd + Z does not revert the changes...
